src/main/java/com/github/sdonkov/PrefixTree.java

```diff
@@ -75,7 +80,115 @@ public boolean contains(CharSequence word) {
         return false;
     }
 
-    public boolean add(CharSequence word) {
+    @Override
+    public int size() {
+        return countNodes(root);
+    }
+
+    private int countNodes(Node node) {
+        int count = node.isEndOfWord() ? 1 : 0;
+        for (Node child : node.edges) {
+            count += countNodes(child);
+        }
+        return count;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (!(o instanceof CharSequence)) {
+            return false;
+        }
+        CharSequence word = (CharSequence) o;
+        TraverseResult traverseResult = traverse((T) word);
+        if (traverseResult.index == word.length() && traverseResult.node.isEndOfWord()) {
+            LOGGER.trace("Tree contains this word - {}", word);
+            return true;
+        }
+        LOGGER.trace("Tree doesn't contain this word - {}", word);
+        return false;
+    }
+
+    public Iterator<T> iterator() {
+        return new Iterator<T>() {
+            private final ArrayList<T> words = new ArrayList<>();
```

Comment:
From what I can gather this collection contains all the words upon instantiation of the iterator.
The point of the iterator is that you are traversing the collection item by item and you can give up at any moment. Ideally, the memory and time requirements to get each item should be constant.

It should be fairly straightforward to implement this with nothing but a crumbpath back to root and an iterator for each of the nodes that branches to more than one sibling. The memory requirement then would be a function of the depth of the tree, not of its size (which is orders of magnitute larger than depth for the use-case a trie is used for).